### ISSUE TITLE
Sometimes missing tnx variable in log entries

### ISSUE THREAD

**Issue by xcdr** [haproxy/haproxy#2304]
### Detailed Description of the Problem

I found a strange problem with the behavior of the haproxy log, it seems to lose the value of the txn variable for some requests.

Tested haproxy  2.6, 2.7, 2.8

Example log entries:
```
2023-09-28T21:13:07+02:00 10.123.140.23 haproxy-lab[2737938]: example-site1 203.0.113.1:58517 [28/Sep/2023:21:13:07.988] main~ varnish/nginx1 0/0/0/1/1 404 532 - - ---- 1/1/0/0/0 0/0 {site.example.org|like Gecko) Chrome/117.0.0.0 Safari/537.36|} "GET /media/cache/Banner_1704x740.jpg HTTP/1.1"
2023-09-28T21:13:01+02:00 10.123.140.23 haproxy-lab[2737938]: default       203.0.113.1:58517 [28/Sep/2023:21:13:01.473] main~ varnish/nginx1 0/0/0/1/1 404 532 - - ---- 1/1/0/0/0 0/0 {site.example.org|like Gecko) Chrome/117.0.0.0 Safari/537.36|} "GET /media/cache/Banner_1704x740.jpg HTTP/1.1"

```
	
When I make curl or single browser request, variable txn.site_code always exists with value based on map() function. 

When browser make many different requests, txn.site_code have "default" value but always proper backend is selected, and backend application sees proper value of X-Site-Code header, never hits the "dummy" backend as it should with "default" value.

### Expected Behavior

Always proper value of txn.site_code in log entries.

### Steps to Reproduce the Behavior

I don't know...

Maybe similar to:
1. do curl requests or single resource from browser for proper log entries
2. do many requests for complete site for inproper log entries

### What is your configuration?

```haproxy
global
    presetenv IP 0.0.0.0

    # Logging - begin
    log /dev/log            local1 notice
    log 10.0.0.1:514 	    local0
    log-tag                 haproxy-lab
    # Logging - end

    chroot /var/lib/haproxy
    user haproxy
    group haproxy
    daemon

...

    stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
    stats timeout 30s

    tune.ssl.default-dh-param 2048

    lua-load /etc/haproxy/lua/robots.lua

defaults
    mode http

    log global
    option dontlognull

    retries 3
    option redispatch
    maxconn 90000 # default ulimit -n

    timeout http-keep-alive 300s

    timeout connect 5s
    timeout client 1m
    timeout server 2m

    option accept-invalid-http-response	

# simplified example frontend
frontend main
    mode http
	
    bind ${IP}:80
    bind ${IP}:443 ssl crt /etc/haproxy/certs/ alpn h2,http/1.1
	
    option httplog

    log-format "%[var(txn.site_code)] %ci:%cp [%tr] %ft %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs %{+Q}r"

    http-request capture req.hdr(Host) len 100
    http-request capture req.hdr(User-Agent) len 100

    http-request set-var(txn.site_code) hdr(host),map(/etc/haproxy/site_code.map,"default")
    http-request set-header X-Site-Code %[var(txn.site_code)]

    acl site_code_unknown var(txn.site_code) -m str "default"

    # Override backends
    use_backend nginx if { var(txn.site_code) -m str example-site } { path_beg -i /foo }


    use_backend %[var(txn.site_code),map(/etc/haproxy/backends.map,dummy)]

backend dummy
    http-request return status 418 content-type "text/plain" string "OK"	
	
backend varnish
...

backend nginx
...
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.8.3-1~bpo11+1 2023/09/08 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2028.
Known bugs: http://www.haproxy.org/bugs/bugs-2.8.3.html
Running on: Linux 5.10.0-21-amd64 #1 SMP Debian 5.10.162-1 (2023-01-21) x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=2).
Built with OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
Running on OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

**Comment by wtarreau** [comment]
What I suspect in fact is that some requests have a `host` header field in upper case or with a port number etc, anything that doesn't match your map, and that you think the request is properly routed to the correct backend because it just happens to be the same as the default one. But given that the same variable is used both for the log and the backend selection, it is the same in both cases.

**Comment by wtarreau** [comment]
Hmmm I'm seeing that it is captured and seems to appear similar in the capture. Strange.

**Comment by wtarreau** [comment]
Since the log was edited to show "site.example.org", are you absolutely certain they were strictly identical, character-for-character, with exactly the same case ? I would suggest that you turn the host hedaer to lower case before mapping it:
```
  http-request set-var(txn.site_code) req.hdr(host),lower,map(/etc/haproxy/site_code.map,"default")
```
At least it would make sure that you properly look up upper cased host names.

**Comment by xcdr** [comment]
> Since the log was edited to show "site.example.org", are you absolutely certain they were strictly identical, character-for-character, with exactly the same case ? I would suggest that you turn the host hedaer to lower case before mapping it:
> 
> ```
>   http-request set-var(txn.site_code) req.hdr(host),lower,map(/etc/haproxy/site_code.map,"default")
> ```
> 
> At least it would make sure that you properly look up upper cased host names.

I tried lowering hostname but it's not problem, map() always works and proper backend selected/used, problem only occurs at log level.

**Comment by wtarreau** [comment]
And you have nothing in other rules (backend etc) which could overwrite txn.side_code ? I'm sorry but given that variables are released *after* the log, I fail to see how this can happen by default.

**Comment by xcdr** [comment]
> And you have nothing in other rules (backend etc) which could overwrite txn.side_code ? I'm sorry but given that variables are released _after_ the log, I fail to see how this can happen by default.

This is weird, because:
- problem occurs on many haproxy instances with different domains 
- this rule works seamless: `use_backend %[var(txn.site_code),map(/etc/haproxy/backends.map,dummy)]` - as You see in logs, dummy backend is not used when txn.site_code has "default" value in log entry (not used in backends.map)
- also works: `http-request set-header X-Site-Code %[var(txn.site_code)]`
- use_backend is last rule, any overwrite of txn.site_code would cause it not to use backend mapped earlier
- if there was no log, I would never have noticed the problem (there are many backends)

**Comment by wtarreau** [comment]
Just to be sure, are you certain that no rule in a backend and no http-response rule ever touches this variable (even by accident after a failed copy-paste for example) ? Also do you have a rough estimate of the ratio of such failures ? E.g. 1/100, 1/10000, 1 in a million, other ?

It could also be useful to check if the affected client addresses experienced any failure in the surrounding requests (just before or just after). This may help us figure if there could be any interaction with failed pipelined request for example.

**Comment by xcdr** [comment]
> Just to be sure, are you certain that no rule in a backend and no http-response rule ever touches this variable (even by accident after a failed copy-paste for example) ? Also do you have a rough estimate of the ratio of such failures ? E.g. 1/100, 1/10000, 1 in a million, other ?

I'am sure there is no other rule that overwrite txn.site_code.

Fast and simple check for log file:
```
# grep 'GET https://example.net/common/images/favicon/apple-touch-icon-76x76.png?v=1' haproxy-lab.log | awk '{ if ($4 != "default") { print $0 } }' | wc -l
16544

# grep 'GET https://example.net/common/images/favicon/apple-touch-icon-76x76.png?v=1' haproxy-lab.log | awk '{ if ($4 = "default") { print $0 } }' | wc -l
16578
```
> It could also be useful to check if the affected client addresses experienced any failure in the surrounding requests (just before or just after). This may help us figure if there could be any interaction with failed pipelined request for example.

I can't see any differences in response codes or counters '/x/y/z/' in haproxy log.

I filtered single IP and no difference in failrue ratio.

Checked traffic via CloudFlare/Akamai/without CDN and via internal LAN frontend - the same result.

**Comment by Darlelet** [comment]
Hi @xcdr, I noticed that you also use lua (from the provided config snippet).
Do you know what `robots.lua` script is doing?

**Comment by wtarreau** [comment]
So it's 50/50 in your case ?

Also there's something odd. The variable is not lost. You're logging the variable itself, and it's either "default" or the name you want. "default" can never be the value of an empty or lost variable. The only explanation I can find for this "default" to appear there is that the variable contains "default" because the entry was not found in the map. Whether it's the map that's gone wild during a lookup or anything is another story, but the fact is that "default" was logged hence it was the variable's content.

**Comment by xcdr** [comment]
> Also there's something odd. The variable is not lost. You're logging the variable itself, and it's either "default" or the name you want. "default" can never be the value of an empty or lost variable. The only explanation I can find for this "default" to appear there is that the variable contains "default" because the entry was not found in the map. Whether it's the map that's gone wild during a lookup or anything is another story, but the fact is that "default" was logged hence it was the variable's content.

"default" is my value for unmapped domain :-)

**Comment by xcdr** [comment]
> Hi @xcdr, I noticed that you also use lua (from the provided config snippet). Do you know what `robots.lua` script is doing?

```
robots = function(applet)
    local response = "User-agent: *\nDisallow: /"
    applet:add_header("Content-Length", string.len(response))
    applet:add_header("Content-Type", "text/plain")
    applet:set_status(200)
    applet:start_response()
    applet:send(response)
end

core.register_service("robots", "http", robots)
```

This service is used only for some domains and path_end /robots.txt

**Comment by xcdr** [comment]
@wtarreau I see something weird, grep for response with code 418 from dummy backend has vary length of response, but this backend return only static "OK" value and I expect constant length, and not expect for backend.

I talked with the tester and he claims that manual and cypress tests never catch 418 except known cases.

**Comment by wtarreau** [comment]
> "default" is my value for unmapped domain :-)

Yes I noticed, what I mean is that it is printed in the logs from the variable, so the variable contains this value, proving that it does exist.

> grep for response with code 418 from dummy backend has vary length of response

be careful, the reported size includes the header encoding, and this one may very likely vary between responses based on those from the client.

**Comment by xcdr** [comment]
> > "default" is my value for unmapped domain :-)
> 
> Yes I noticed, what I mean is that it is printed in the logs from the variable, so the variable contains this value, proving that it does exist.
> 

True, but for the same site/path should have always the same value.

> > grep for response with code 418 from dummy backend has vary length of response
> 
> be careful, the reported size includes the header encoding, and this one may very likely vary between responses based on those from the client.

You are right about length, but HTTP 418 never should be in logs, looks similar that default/example-site1 value of txn.site_code :-/

**Comment by wtarreau** [comment]
Which further reinforces my initial analysis that the variable doesn't change and that the request is routed to the wrong backend (the one being mapped to "defaults") and/or that the backend server that extracts the header "X-Site-Code" routes somewhere unexpected on "defaults" and that this "somewhere" emits the 418. If you're using clear text commnications between haproxy and your servers, you should definitely start a tcpdump for a while so that you can later check where these "default" request go and who responds 418. Since this code was assigned as an April's fool joke, it very likely comes from a library or middleware somewhere down the chain, possibly for unhandled x-site-code!

**Comment by xcdr** [comment]
> Which further reinforces my initial analysis that the variable doesn't change and that the request is routed to the wrong backend (the one being mapped to "defaults") and/or that the backend server that extracts the header "X-Site-Code" routes somewhere unexpected on "defaults" and that this "somewhere" emits the 418. If you're using clear text commnications between haproxy and your servers, you should definitely start a tcpdump for a while so that you can later check where these "default" request go and who responds 418. Since this code was assigned as an April's fool joke, it very likely comes from a library or middleware somewhere down the chain, possibly for unhandled x-site-code!

No, no,  wrong way :-)

look carefully to the included configuration and what I wrote.

This is weird because configuration seems work OK and problem occurs only in log entries (read as: 0==1...):
* 418 i response from my "dummy" backend in configuration that I included, occurs only in logs but not in real http response for this requests (confirmed by many people)
* The same request has x-site-code=example-site1 captured in the backend app when txn.site_code in log file has "default" value

UPDATE: I updated included config, my mistake in sample config that I posted:
before: `http-request return status 200 content-type "text/plain" string "OK"`
after: `http-request return status 418 content-type "text/plain" string "OK"`

**Comment by xcdr** [comment]
@wtarreau  I found something interesting, I changed config like that:

```
# added logging of txn.site_code -> proc.site_code
log-format "%[var(txn.site_code)] -> %[var(proc.site_code)] %ci:%cp [%tr] %ft %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs %{+Q}r"
```

```
# added paraller mapping of proc.site_code
http-request set-var(txn.site_code) hdr(host),map(/etc/haproxy/site_code.map,"default")
http-request set-var(proc.site_code) hdr(host),map(/etc/haproxy/site_code.map,"default")

http-request set-header X-Site-Code %[var(txn.site_code)]
```

Now my logs looks similar to:
`2023-09-28T21:13:07+02:00 10.123.140.23 haproxy-lab[2737938]: default -> example-site1 203.0.113.1:58517 [28/Sep/2023:21:13:07.988] main~ varnish/nginx1 0/0/0/1/1 404 532 - - ---- 1/1/0/0/0 0/0 {site.example.org|like Gecko) Chrome/117.0.0.0 Safari/537.36|} "GET /media/cache/Banner_1704x740.jpg HTTP/1.1"`

Additionally I captured via varnishlog one of backend and confirmed that the same request has header: x-site-code=example-site1

Why logged value of txn.site_code is different from proc.site_code?

**Comment by xcdr** [comment]
@wtarreau THIS IT NOT haproxy BUG, this is my stupid mistake :-)

After many steps with log analysis I lost "=" from awk comparison!

Should be:  `awk '{ if ($4 == "default") { print $0 } }'`,  but ater many test, changes column, operators !=, ==, ~, I ended with: `awk '{ if ($4 = "default") { print $0 } }'`.

**Comment by wtarreau** [comment]
Ah, excellent! Kudos for spotting this one! At least it makes much more sense now. Many thanks for reporting it, some would have discretely abandoned the ticket :-)

Now to respond to your other question regarding txn.side_code vs proc.site_code, keep in mind that there will be one txn variable per request in flight, while there will be a single global proc variable, so when using proc, the last one which writes wins, hence it was not suitable for your use case (but I understand how it might have helped you debug it).

**Comment by xcdr** [comment]
Thank You @wtarreau, and sorry for confusing ticket.

On my side this is "Fog of War", many instances of haproxy/varnish/nginx with huge config (included above is only essential example), additional load balances (forti/akamai/cloudflare) in front of haproxy.... and finally stupid typo causes mind madness for many hours.
